man/fax2ps.1
.\"	$Header: /usr/people/sam/fax/./man/RCS/fax2ps.1,v 1.13 1995/03/17 18:52:58 sam Rel $
.\"
.\" Copyright (c) 1991, 1992, 1993, 1994, 1995 by Sam Leffler.
.\" All rights reserved.
.\"
.\" This file is provided for unrestricted use provided that this
.\" legend is included on all tape media and as a part of the
.\" software program in whole or part.  Users may copy, modify or
.\" distribute this file at will.
.\"
.if n .po 0
.ds Ps P\s-2OST\s+2S\s-2CRIPT\s+2
.TH FAX2PS 1 "March 16, 1995"
.SH NAME
fax2ps \- convert a
.SM TIFF
facsimile to compressed \*(Ps\(tm
.SH SYNOPSIS
.B fax2ps
[
.I options
] [
.IR file ...
]
.SH DESCRIPTION
.I fax2ps
reads one or more
.SM TIFF
facsimile image files and prints a compressed form of
\*(Ps on the standard output that is suitable for printing.
.PP
By default, each page is scaled to reflect the
image dimensions and resolutions stored in the file.
The
.B \-x
and
.B \-y
options can be used to specify the horizontal and vertical
image resolutions (lines/inch), respectively.
If the
.B \-S
option is specified, each page is scaled to fill an output page.
The default output page is 8.5 by 11 inches.
Alternate page dimensions can be specified in inches with the
.B \-W
and
.B \-H
options.
.PP
By default
.I fax2ps
generates \*(Ps for all pages in the file.
The
.B \-p
option can be used to select one or more pages from
a multi-page document.
.PP
.I fax2ps
generates a compressed form of \*(Ps that is
optimized for sending pages of text to a \*(Ps
printer attached to a host through a low-speed link (such
as a serial line).
Each output page is filled with white and then only
the black areas are drawn.
The \*(Ps specification of the black drawing operations
is optimized by using a special font that encodes the
move-draw operations required to fill
the black regions on the page.
This compression scheme typically results in a substantially
reduced \*(Ps description, relative to the straightforward
imaging of the page with a \*(Ps
.I image
operator.
This algorithm can, however, be ineffective
for continuous-tone and white-on-black images.
For these images, it sometimes is more efficient to send
the raster bitmap image directly; see
.IR tiff2ps (1).
.SH OPTIONS
.TP 10
.BI \-p " number"
Print only the indicated page.
Multiple pages may be printed by specifying
this option more than once.
.TP 10
.BI \-x " resolution"
Use
.I resolution
as the horizontal resolution, in dots/inch, of the image data.
By default this value is taken from the file.
.TP 10
.BI \-y " resolution"
Use
.I resolution
as the vertical resolution, in lines/inch, of the image data.
By default this value is taken from the file.
.TP 10
.B \-S
Scale each page of image data to fill the output page dimensions.
By default images are presented according to the dimension
information recorded in the 
.SM TIFF
file.
.TP 10
.BI \-W " width"
Use
.I width
as the width, in inches, of the output page.
The default page width is 8.5 inches.
.TP 10
.BI \-H " height"
Use
.I height
as the height, in inches, of the output page.
The default page height is 11 inches.
.SH DIAGNOSTICS
Some messages about malformed 
.SM TIFF
images come from the
.SM TIFF
library.
.PP
Various messages about badly formatted facsimile images
may be generated due to transmission errors in received
facsimile.
.I fax2ps
attempts to recover from such data errors by resynchronizing
decoding at the end of the current scanline.
This can result in long horizontal black lines in the resultant
\*(Ps image.
.SH NOTES
If the destination printer supports \*(Ps Level II then
it is always faster to just send the encoded bitmap generated
by the
.IR tiff2ps (1)
program.
.SH BUGS
.I fax2ps
should probably figure out when it is doing a poor
job of compressing the output and just generate 
\*(Ps to image the bitmap raster instead.
.SH "SEE ALSO"
.IR tiff2ps (1),
.IR libtiff (3)


CHANGES
#	$Id: CHANGES 1078 2012-01-23 19:02:31Z faxguy $	

New Changes

* prevent more problems due to corrupt sendq/doneq files (23 Jan 2012)

(5.5.1)

* fix some information presented by faxwatch (2 Jan 2012)
* fix JPARM JOBTYPE parameter (27 Dec 2011)
* remove useless MAXPAGES hfaxd JPARM option (27 Dec 2011)
* add support for proprietary LSI +FHS: codes 12-16 (6 Dec 2011)
* fix %r for ModemFmt (28 Nov 2011)
* add faxq Proxy feature through JobControl (28 Nov, 13, 21 Dec 2011)
* add duration and conntime job details to sendq files (28 Nov 2011)
* add -N option to faxalter (28 Nov 2011)
* make hfaxd more firewall-friendly (3 Nov 2011)
* fix color sending support for A4, B4, and A3 (2 Nov 2011)
* add support for the MT9234ZBA (12 Sep 2011)
* add recursive dialrules matching (11 Sep 2011)
* fix zero-length dialrules matching (11 Sep 2011)
* disable AUTOROTATE by default for old Ghostscript (5 Sep 2011)
* disable unnecessary debugging in hfaxd pasv mode (3 Sep 2011)
* fix various problems revealed by cppcheck (3 Sep 2011)
* fix problem with octet-stream MIME encoder (23 Aug 2011)
* prevent a faxq crash due to a corrupt sendq file (22 Aug 2011)
* remove -dFIXEDMEDIA from FIXEDWIDTH, again (24 May 2011)
* fix job-based TimeOfDay operation (5 Mar 2011)
* add Russian and Ukranian dictionary translations (16 Feb 2011)

(5.5.0)

* fix typerules byte type offset handling (4 Dec 2010)
* fix stripbytecount problem in libtiff 4.0 support (3 Dec 2010)
* use -dEPSFitPage -dPDFFitPage -dFIXEDMEDIA by default (29 Nov 2010)
* revert to pswrite for faxmail pdf MIMEConverter (27 Nov 2010)
* replace 'wc -L' with awk in scripting for portability (23 Nov 2010)
* add faxmail FirstAlternativeOnly configuration option and only
  use the first alternative part by default (22 Nov 2010)
* add support for UTF-8 in textfmt and faxmail (22 Nov 2010)
* detect hasV34Trouble for Class 1.0 DLE+EOT dial responses (20 Nov 2010)
* fix scheduler throttle (17 Nov 2010)
* add MaxTraversal scheduler config option (16 Nov 2010)
* use tiff2pdf instead of tiff2ps in tiff2fax conversions (6 Nov 2010)
* enable auto-rotate.ps by default with AUTOROTATE FaxModify feature
  (4 Nov 2010)
* make FIXEDWIDTH=-dPDFFitPage default instead of -dFIXEDMEDIA (4 Nov 2010)
* wait for OK at the end of a Class 1.0 V.34 disconnection (4 Nov 2010)
* expand color page buffer size to 2MB (25 Oct 2010)
* check for REG_STARTEND define for system regex support (25 Oct 2010)
* improve auto-rotate.ps to match most page sizes (25 Oct 2010)

(5.4.3)

* add faxadduser -c option (16 Oct 2010)
* detect hasV34Trouble for Class 1.0 "OK" dial responses (8 Oct 2010)
* fix build for gcc-4.5 (25 Sep 2010)
* use system regex library by default instead (20 Sep 2010)
* in faxmail's PDF MIMEConverter scale non-letter images onto letter 
  pages instead of cutting off parts of them (14 Sep 2010)
* in Class 1 sending ECM after CTC/CTR adjust dataTimeout to new bitrate
  (26 Jul 2010)
* add Class1ECMCheckFrameLength configuration option (24 Jun, 29 Aug 2010)
* improve the faxmail quoted-printable boundary check (19 Jun 2010)
* Use -dMaxStripSize=0 with Ghostscript calls (6 May 2010)

(5.4.2)

* work around a Ghostscript 8.71 segfault on genfontmap.ps (26 Apr 2010)
* cope with multi-strip MMR images prepared by Ghostscript (26 Apr 2010)
* handle unexpected receipt of RR when the sender didn't hear our MCF
  (25 Apr 2010)
* use ps2write device by default if available for faxmail pdf MIMEConverter
  (25 Apr 2010)
* cope with some modems' broken DTMF reporting when possible (23 Mar 2010)
* don't add a whitespace before typerules %o or %i (21 Mar 2010)

(5.4.1)

* be consistent to use libtiff tools from TIFFBIN in scripts (11 Feb 2010)
* add an alternative to open_memstream and fmemopen for systems 
  without these functions (8 Feb 2010)
* fix a problem with the calculation of the page data size which was 
  introduced in 5.4.0 (7 Feb 2010)
* add MaxConcurrentPreps faxq config feature and default to 1 (6 Feb 2010)
* fix a problem with the ANSWER "any" hfaxd command (4 Feb 2010)
* add ModemAnswerAgainCmd modem configuration option (20 Jan 2010)
* scan the entire page of image data for RTC when scanning for it (18 Jan 2010)
* wait longer for Phase C carrier drop in Class 1 (18 Jan 2010)
* stop decoding non-ECM page image data as soon as RTC is seen (18 Jan 2010)
* fix a potential hang in sending Class 1 data if the modem is hung (16 Jan 2010)
* fix faxdeluser from changing the gid of hosts.hfaxd (2 Jan 2010)
* change faxadduser to add new entries to the top of hosts.hfaxd (2 Jan 2010)
* in Class 1 if CONNECT is unexpectely received, it's an error (30 Dec 2009)

(5.4.0)

* add support for sending color faxes (8, 21 Dec 2009)
* use genfontmap.ps to create Fontmap.HylaFAX from faxsetup (5 Dec 2009)
* add LogSend and LogRecv modem configuration options (3 Dec 2009)
* use LittleCMS to convert received color faxes from ITULAB (3 Dec 2009)
* add Flemish dictionary translation (2 Dec 2009)
* handle echo of PPM when sending (9 Nov 2009)
* add support for libtiff v4.0 (6, 9 Nov 2009)
* faxmail: handle wrapped headers better (28 Oct 2009)

(5.3.0)

* add config option (usually -O) to the clients (29 Sep 2009)
* add -c option for hfaxd and faxq (29 Sep 2009)
* properly unblock jobs when job preparation fails (24 Sep 2009)
* make PostscriptTimeout format failures visible (24 Sep 2009)
* add auto-rotate.ps utility script (7 Sep 2009)
* add dialrules-pabx.be sample (7 Sep 2009)
* add "bindaddressfamily" option to hfaxd and default to IPv4 (7 Sep 2009)
* support Ghostscript 8.64 font aliases (28 Jul 2009)
* handle echo of PPS when sending (29 Apr 2009)
* improve detection of disabled Mac OS X (Darwin) realtime scheduling (21 Mar 2009)
* add IPv6 support (1 Mar 2009)

(5.2.9)

* cope better in cases where the modem may "ignore" our ATA (16 Feb 2009)
* fix problem involving JBIG attempts using non-JBIG Class 2.0 modems
  (20 Jan 2009)

fax2ps/fax2ps.c
/* $Header: /usr/people/sam/tiff/contrib/fax2ps/RCS/fax2ps.c,v 1.33.2.1 1995/03/16 18:56:30 sam Exp $" */

/*
 * Copyright (c) 1991, 1992, 1993, 1994 by Sam Leffler.
 * All rights reserved.
 *
 * This file is provided for unrestricted use provided that this
 * legend is included on all tape media and as a part of the
 * software program in whole or part.  Users may copy, modify or
 * distribute this file at will.
 */
#include <math.h>
#include <stdio.h>
#include <string.h>

#include "defs.h"

float	defxres = 204.;		/* default x resolution (pixels/inch) */
float	defyres = 98.;		/* default y resolution (lines/inch) */
float	pageWidth = 8.5;	/* image page width (inches) */
float	pageHeight = 11.0;	/* image page length (inches) */
int	scaleToPage = FALSE;	/* if true, scale raster to page dimensions */

static	void do_font_insert(void);

static void
setupPass(TIFF* tif, int pass)
{
    uint32* stripbytecount;

    fax.pass = pass;
    fax.row = 0;
    TIFFGetField(tif, TIFFTAG_STRIPBYTECOUNTS, &stripbytecount);
    fax.cc = stripbytecount[0];
    fax.bp = fax.buf;
    FaxPreDecode(tif);
}

static	int totalPages = 0;

void
printTIF(TIFF* tif, int pageNumber)
{
    uint32 w, h;
    uint16 fill, unit, photometric, compression;
    float xres, yres;
    uint32 g3opts;
    uint32* stripbytecount;

    TIFFGetField(tif, TIFFTAG_STRIPBYTECOUNTS, &stripbytecount);
    fax.cc = stripbytecount[0];
    fax.buf = (u_char*) malloc(fax.cc);
    TIFFReadRawStrip(tif, 0, fax.buf, fax.cc);
    if (TIFFGetField(tif,TIFFTAG_FILLORDER, &fill) && fill != FILLORDER_MSB2LSB)
	TIFFReverseBits(fax.buf, fax.cc);
    TIFFGetField(tif, TIFFTAG_IMAGELENGTH, &h);
    TIFFGetField(tif, TIFFTAG_IMAGEWIDTH, &w);
    if (!TIFFGetField(tif, TIFFTAG_XRESOLUTION, &xres)) {
	TIFFWarning(TIFFFileName(tif),
	    "No x-resolution, assuming %g dpi", defxres);
	xres = defxres;
    }
    if (!TIFFGetField(tif, TIFFTAG_YRESOLUTION, &yres)) {
	TIFFWarning(TIFFFileName(tif),
	    "No y-resolution, assuming %g lpi", defyres);
	yres = defyres;					/* XXX */
    }
    if (TIFFGetField(tif, TIFFTAG_RESOLUTIONUNIT, &unit) &&
      unit == RESUNIT_CENTIMETER) {
	xres *= 25.4;
	yres *= 25.4;
    }
    TIFFGetField(tif, TIFFTAG_PHOTOMETRIC, &photometric);
    /*
     * Calculate the scanline/tile widths.
     */
    if (isTiled(tif)) {
        fax.b.rowbytes = TIFFTileRowSize(tif);
	TIFFGetField(tif, TIFFTAG_TILEWIDTH, &fax.b.rowpixels);
    } else {
        fax.b.rowbytes = TIFFScanlineSize(tif);
	TIFFGetField(tif, TIFFTAG_IMAGEWIDTH, &fax.b.rowpixels);
    }
    if (TIFFGetField(tif, TIFFTAG_GROUP3OPTIONS, &g3opts))
	fax.is2d = (g3opts & GROUP3OPT_2DENCODING) != 0;
    else
	fax.is2d = 0;
    fax.scanline = (u_char*) malloc(2*fax.b.rowbytes);
    fax.b.refline = fax.scanline + fax.b.rowbytes;
    TIFFGetField(tif, TIFFTAG_COMPRESSION, &compression);
    if (compression == COMPRESSION_CCITTFAX4)
	fax.options = FAX3_NOEOL;

    printf("%%%%Page: \"%d\" %d\n", pageNumber, pageNumber);
    printf("gsave\n");
    if (scaleToPage) {
	float yscale = pageHeight / (h/yres);
	float xscale = pageWidth / (w/xres);
	printf("0 %d translate\n", (int)(yscale*(h/yres)*72.));
	printf("%g %g scale\n", (72.*xscale)/xres, -(72.*yscale)/yres);
    } else {
	printf("0 %d translate\n", (int)(72.*h/yres));
	printf("%g %g scale\n", 72./xres, -72./yres);
    }
    printf("0 setgray\n");
    setupPass(tif, 3);
    while (fax.cc > 0) {
	if (compression == COMPRESSION_CCITTFAX4)
	    Fax4DecodeRow(tif, w);
	else
	    Fax3DecodeRow(tif, w);
	print_scan_row(w, fax.scanline);
    }
    printf("p\n");
    printf("grestore\n");
    free((char*) fax.scanline);
    free((char*) fax.buf);
    totalPages++;
}

#define	GetPageNumber(tif) \
TIFFGetField(tif, TIFFTAG_PAGENUMBER, &pn, &ptotal)

int
findPage(TIFF* tif, int pageNumber)
{
    uint16 pn = (uint16) -1;
    uint16 ptotal = (uint16) -1;
    if (GetPageNumber(tif)) {
	while (pn != pageNumber && TIFFReadDirectory(tif) && GetPageNumber(tif))
	    ;
	return (pn == pageNumber);
    } else
	return (TIFFSetDirectory(tif, pageNumber-1));
}

void
fax2ps(TIFF* tif, int npages, int* pages, char* filename)
{
    if (npages > 0) {
	uint16 pn, ptotal;
	int i;

	if (!GetPageNumber(tif))
	    fprintf(stderr, "%s: No page numbers, counting directories.\n",
		filename);
	for (i = 0; i < npages; i++) {
	    if (findPage(tif, pages[i]))
		printTIF(tif, pages[i]);
	    else
		fprintf(stderr, "%s: No page number %d\n", filename, pages[i]);
	}
    } else {
	int pageNumber = 1;
	do
	    printTIF(tif, pageNumber++);
	while (TIFFReadDirectory(tif));
    }
}

#undef GetPageNumber

static int
pcompar(void* va, void* vb)
{
    int* pa = (int*) va;
    int* pb = (int*) vb;
    return (*pa - *pb);
}

extern	double atof();

main(int argc, char** argv)
{
    extern int optind;
    extern char* optarg;
    int c, pageNumber;
    int* pages = 0, npages = 0;
    int dowarnings = FALSE;	/* if 1, enable library warnings */
    long t;
    TIFF* tif;

    while ((c = getopt(argc, argv, "p:x:y:W:H:wS")) != -1)
	switch (c) {
	case 'H':		/* page height */
	    pageHeight = atof(optarg);
	    break;
	case 'S':		/* scale to page */
	    scaleToPage = TRUE;
	    break;
	case 'W':		/* page width */
	    pageWidth = atof(optarg);
	    break;
	case 'p':		/* print specific page */
	    pageNumber = atoi(optarg);
	    if (pageNumber < 1) {
		fprintf(stderr, "%s: Invalid page number (must be > 0).\n",
		    optarg);
		exit(-1);
	    }
	    if (pages)
		pages = (int*) realloc((char*) pages, (npages+1)*sizeof (int));
	    else
		pages = (int*) malloc(sizeof (int));
	    pages[npages++] = pageNumber;
	    break;
	case 'w':
	    dowarnings = TRUE;
	    break;
	case 'x':
	    defxres = atof(optarg);
	    break;
	case 'y':
	    defyres = atof(optarg);
	    break;
	case '?':
	    fprintf(stderr,
"usage: %s [-w] [-p pagenumber] [-x xres] [-y res] [-S] [-W pagewidth] [-H pageheight] [files]\n",
		argv[0]);
	    exit(-1);
	}
    if (npages > 0)
	qsort(pages, npages, sizeof (int), pcompar);
    if (!dowarnings)
	TIFFSetWarningHandler(0);
    printf("%%!PS-Adobe-3.0\n");
    printf("%%%%Creator: fax2ps\n");
#ifdef notdef
    printf("%%%%Title: %s\n", file);
#endif
    t = time(0);
    printf("%%%%CreationDate: %s", ctime(&t));
    printf("%%%%Origin: 0 0\n");
    printf("%%%%BoundingBox: 0 0 %u %u\n",
	(int)(pageHeight*72), (int)(pageWidth*72));	/* XXX */
    printf("%%%%Pages: (atend)\n");
    printf("%%%%EndComments\n");
    printf("%%%%BeginProlog\n");
    do_font_insert();
    printf("/d{bind def}def\n"); /* bind and def proc */
    printf("/m{0 exch moveto}d\n");
    printf("/s{show}d\n");
    printf("/p{showpage}d \n");	/* end page */
    printf("%%%%EndProlog\n");
    if (optind < argc) {
	do {
	    tif = TIFFOpen(argv[optind], "r");
	    if (tif) {
		fax2ps(tif, npages, pages, argv[optind]);
		TIFFClose(tif);
	    } else
		fprintf(stderr, "%s: Can not open, or not a TIFF file.\n",
		    argv[optind]);
	} while (++optind < argc);
    } else {
	int n, fd;
	char temp[1024], buf[16*1024];

	strcpy(temp, "/tmp/fax2psXXXXXX");
	fd = mkstemp(temp);
	if (fd == -1) {
	    fprintf(stderr, "Could not create temp file \"%s\"\n", temp);
	    exit(-2);
	}
	while ((n = read(fileno(stdin), buf, sizeof (buf))) > 0)
	    write(fd, buf, n);
	tif = TIFFOpen(temp, "r");
	unlink(temp);
	if (tif) {
	    fax2ps(tif, npages, pages, "<stdin>");
	    TIFFClose(tif);
	} else
	    fprintf(stderr, "%s: Can not open, or not a TIFF file.\n", temp);
	close(fd);
    }
    printf("%%%%Trailer\n");
    printf("%%%%Pages: %u\n", totalPages);
    printf("%%%%EOF\n");

    exit(0);
}

/* 
 * Create a special PostScript font for printing FAX documents.  By taking
 * advantage of the font-cacheing mechanism, a substantial speed-up in 
 * rendering time is realized. 
 */
static void
do_font_insert(void)
{
    puts("/newfont 10 dict def newfont begin /FontType 3 def /FontMatrix [1");
    puts("0 0 1 0 0] def /FontBBox [0 0 512 1] def /Encoding 256 array def");
    puts("0 1 31{Encoding exch /255 put}for 120 1 255{Encoding exch /255");
    puts("put}for Encoding 37 /255 put Encoding 40 /255 put Encoding 41 /255");
    puts("put Encoding 92 /255 put /count 0 def /ls{Encoding exch count 3");
    puts("string cvs cvn put /count count 1 add def}def 32 1 36{ls}for");
    puts("38 1 39{ls}for 42 1 91{ls}for 93 1 99{ls}for /count 100");
    puts("def 100 1 119{ls}for /CharDict 5 dict def CharDict begin /white");
    puts("{dup 255 eq{pop}{1 dict begin 100 sub neg 512 exch bitshift");
    puts("/cw exch def cw 0 0 0 cw 1 setcachedevice end}ifelse}def /black");
    puts("{dup 255 eq{pop}{1 dict begin 110 sub neg 512 exch bitshift");
    puts("/cw exch def cw 0 0 0 cw 1 setcachedevice 0 0 moveto cw 0 rlineto");
    puts("0 1 rlineto cw neg 0 rlineto closepath fill end}ifelse}def /numbuild");
    puts("{dup 255 eq{pop}{6 0 0 0 6 1 setcachedevice 0 1 5{0 moveto");
    puts("dup 32 and 32 eq{1 0 rlineto 0 1 rlineto -1 0 rlineto closepath");
    puts("fill newpath}if 1 bitshift}for pop}ifelse}def /.notdef {}");
    puts("def /255 {}def end /BuildChar{exch begin dup 110 ge{Encoding");
    puts("exch get 3 string cvs cvi CharDict /black get}{dup 100 ge {Encoding");
    puts("exch get 3 string cvs cvi CharDict /white get}{Encoding exch get");
    puts("3 string cvs cvi CharDict /numbuild get}ifelse}ifelse exec end");
    puts("}def end /Bitfont newfont definefont 1 scalefont setfont");
}
